File: embeddeds/java/src/main/java/io/github/mosser/arduinoml/embedded/java/dsl/demo/MultiStateAlarm.java
package io.github.mosser.arduinoml.embedded.java.dsl.demo;

import io.github.mosser.arduinoml.kernel.App;
import io.github.mosser.arduinoml.kernel.generator.ToWiring;
import io.github.mosser.arduinoml.kernel.generator.Visitor;

import static io.github.mosser.arduinoml.embedded.java.dsl.AppBuilder.actuator;
import static io.github.mosser.arduinoml.embedded.java.dsl.AppBuilder.application;
import static io.github.mosser.arduinoml.embedded.java.dsl.AppBuilder.sensor;

public class MultiStateAlarm {
    public static void main(String[] args) {

        App myApp = application("red_button")
                .uses(sensor("button", 8))
                .uses(actuator("red_led", 12))
                .uses(actuator("buzzer", 9))

                .hasForState("ready").initial()
                .setting("red_led").toLow()
                .setting("buzzer").toLow()
                .endState()

                .hasForState("buzzing")
                .setting("red_led").toLow()
                .setting("buzzer").toHigh()
                .endState()

                .hasForState("led_on")
                .setting("red_led").toHigh()
                .setting("buzzer").toLow()
                .endState()

                .beginTransitionTable()
                .from("ready")
                .when("button").isHigh()
                .goTo("buzzing")

                .from("buzzing")
                .when("button").isHigh()
                .goTo("led_on")

                .from("led_on")
                .when("button").isHigh()
                .goTo("ready")
                .endTransitionTable()

                .build();

        Visitor codeGenerator = new ToWiring();
        myApp.accept(codeGenerator);
        System.out.println(codeGenerator.getResult());
    }
}

File: embeddeds/scala/src/main/scala/io/github/mosser/arduinoml/samples/MultiStateAlarm.scala
package io.github.mosser.arduinoml.samples

import io.github.mosser.arduinoml.dsl.ArduinoML

object MultiStateAlarm extends App with ArduinoML {

  this hasForName "red_button"

  val button = declare aSensor()    named "button"  boundToPin 8
  val redLed = declare anActuator() named "red_led" boundToPin 12
  val buzzer = declare anActuator() named "buzzer"  boundToPin 9

  val ready = state named "ready" executing (
    redLed --> low,
    buzzer --> low
  )

  val buzzing = state named "buzzing" executing (
    redLed --> low,
    buzzer --> high
  )

  val ledOn = state named "led_on" executing (
    redLed --> high,
    buzzer --> low
  )

  ready.isInitial

  transitions {
    ready   -> buzzing when (button is high)
    buzzing -> ledOn  when (button is high)
    ledOn   -> ready  when (button is high)
  }

  exportToWiring
}
